src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -1184,11 +1184,24 @@ private void ExecuteAndCatchIfThrows_NoLock<TArg>(DiagnosticAnalyzer analyzer, A
                 cancellationToken.ThrowIfCancellationRequested();
                 analyze(argument);
             }
-            catch (Exception ex) when (HandleAnalyzerException(ex, analyzer, info, OnAnalyzerException, _analyzerExceptionFilter, cancellationToken))
+            catch (Exception ex) when (HandleAnalyzerException(analyzer, ex, info) &&
+                HandleAnalyzerException(ex, analyzer, info, OnAnalyzerException, _analyzerExceptionFilter, cancellationToken))
             {
             }
         }
 
+        private bool HandleAnalyzerException(DiagnosticAnalyzer analyzer, Exception ex, in AnalysisContextInfo? info)
+        {
+            if (this.Compilation.FeatureDebugAnalyzers)
+            {
+                Debug.Assert(false);
```

Comment:
Do we want to assert in the filter block, or just FailFast?

Reply:
Assert will present you with the nice "Attach debugger" dialog on .NET Framework which can be useful. Plus it might be friendlier that it throws an exception instead of failing fast in Debug config.